pkg/workceptor/kubernetes.go

```diff
@@ -0,0 +1,326 @@
+		}
+		return
+	}
+	err = saveStatus(kw.unitdir, WorkStateRunning, "Pod running", 0)
+	if err != nil {
+		logger.Error("Unable to save status: %s", err)
+	}
+	//TODO: This blocks while the whole command runs.  Need to wrap this in a context somehow.
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  stdin,
+		Stdout: stdout,
+		Tty:    false,
+	})
+	if err != nil {
+		err = saveStatus(kw.unitdir, WorkStateFailed, fmt.Sprintf("Error attaching to pod: %s", err), 0)
+		if err != nil {
+			logger.Error("Unable to save status: %s", err)
+		}
+		return
+	}
+	err = saveStatus(kw.unitdir, WorkStateSucceeded, "Finished", stdout.Size())
+	if err != nil {
+		logger.Error("Unable to save status: %s", err)
+	}
+}
+
+// Start launches a job with given parameters.
+func (kw *kubeUnit) Start(params string, unitdir string) error {
+	kw.unitdir = unitdir
+	err := saveStatus(kw.unitdir, WorkStatePending, "Connecting to Kubernetes", 0)
+	if err != nil {
+		logger.Error("Unable to save status: %s", err)
+	}
+	kw.ctx, kw.cancelFunc = context.WithCancel(context.Background())
+
+	// Figure out command and args
+	var command []string = nil
+	args, err := shlex.Split(params)
+	if err != nil {
+		return err
+	}
+	if len(kw.command) > 0 {
+		command, err = shlex.Split(kw.command[0])
+		if err != nil {
+			return err
+		}
+		command = append(command, args...)
+		args = nil
+	}
+	kw.command = command
+	kw.args = args
+
+	// Set up Kubernetes API connection
+	clr := clientcmd.NewDefaultClientConfigLoadingRules()
+	if kw.kubeConfig != "" {
+		clr.ExplicitPath = kw.kubeConfig
+	}
+	if kw.namespace == "" {
+		c, err := clr.Load()
+		if err != nil {
+			return err
+		}
+		kw.namespace = c.Contexts[c.CurrentContext].Namespace
+	}
+	kw.config, err = clientcmd.BuildConfigFromFlags("", clr.GetDefaultFilename())
+	if err != nil {
+		return err
+	}
+	kw.clientset, err = kubernetes.NewForConfig(kw.config)
+	if err != nil {
+		return err
+	}
+
+	// Launch runner process
+	go kw.runWork()
+
+	return nil
+}
+
+// Cancel releases resources associated with a job, including cancelling it if running.
+func (kw *kubeUnit) Cancel() error {
+	if kw.pod != nil {
+		go func(podName string) {
+			err := kw.clientset.CoreV1().Pods(kw.namespace).Delete(context.Background(), podName, metav1.DeleteOptions{})
+			if err != nil {
+				logger.Error("Error deleting pod %s: %s", podName, err)
+			}
+		}(kw.pod.Name)
+	}
+	kw.cancelFunc()
+	return nil
+}
+
+// **************************************************************************
+// Command line
+// **************************************************************************
+
+// WorkKubeCfg is the cmdline configuration object for a Kubernetes worker plugin
+type WorkKubeCfg struct {
+	Service    string `required:"true" description:"Local Receptor service name to bind to"`
+	KubeConfig string `description:"Kubeconfig file (defaults to environment)"`
+	Project    string `required:"true" description:"Kubernetes project name"`
```

Comment:
Should this be `Namespace`?

Reply:
It's called a Project in all the user-facing screens in the OpenShift GUI, but a Namespace in all the Kubernetes API calls.  Not sure which we would consider more correct here, but I'm happy to change it if you think Namespace is better.

Reply:
I guess nothing else in this is OpenShift specific, and there really isn't such a thing as a "Kubernetes project," so you're right, this should be changed.